dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBMongoDB/AzureCosmosDBMongoDBVectorStoreRecordMapper.cs

```diff
@@ -0,0 +1,152 @@
+    private static readonly HashSet<Type> s_supportedVectorTypes =
+    [
+        typeof(ReadOnlyMemory<float>),
+        typeof(ReadOnlyMemory<float>?),
+        typeof(ReadOnlyMemory<double>),
+        typeof(ReadOnlyMemory<double>?)
+    ];
+
+    /// <summary>A dictionary that maps from a property name to the storage name.</summary>
+    private readonly Dictionary<string, string> _storagePropertyNames = [];
+
+    /// <summary>A dictionary that maps from a storage property name to the data model property name.</summary>
+    private readonly Dictionary<string, string> _reversedStoragePropertyNames = [];
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AzureCosmosDBMongoDBVectorStoreRecordMapper{TRecord}"/> class.
+    /// </summary>
+    /// <param name="keyProperty">A property info object that points at the key property for the current model, allowing easy reading and writing of this property.</param>
+    /// <param name="dataProperties">A list of property info objects that point at the data properties in the current model, and allows easy reading and writing of these properties.</param>
+    /// <param name="vectorProperties">A list of property info objects that point at the vector properties in the current model, and allows easy reading and writing of these properties.</param>
+    /// <param name="storagePropertyNames">A dictionary that maps from a property name to the configured name that should be used when storing it.</param>
+    public AzureCosmosDBMongoDBVectorStoreRecordMapper(PropertyInfo keyProperty, List<PropertyInfo> dataProperties, List<PropertyInfo> vectorProperties, Dictionary<string, string> storagePropertyNames)
+    {
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes([keyProperty], s_supportedKeyTypes, "Key");
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes(dataProperties, s_supportedDataTypes, "Data", supportEnumerable: true);
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes(vectorProperties, s_supportedVectorTypes, "Vector");
+
+        this._storagePropertyNames = storagePropertyNames;
+
+        // Use Mongo reserved key property name as storage key property name
+        this._storagePropertyNames[keyProperty.Name] = AzureCosmosDBMongoDBConstants.MongoReservedKeyPropertyName;
+
+        this._reversedStoragePropertyNames = ReverseDictionary(storagePropertyNames);
+    }
+
+    public BsonDocument MapFromDataToStorageModel(TRecord dataModel)
+        => MapDocument(dataModel.ToBsonDocument(), this._storagePropertyNames);
+
+    public TRecord MapFromStorageToDataModel(BsonDocument storageModel, StorageToDataModelMapperOptions options)
+        => BsonSerializer.Deserialize<TRecord>(MapDocument(storageModel, this._reversedStoragePropertyNames));
+
+    #region private
+
+    private static BsonDocument MapDocument(BsonDocument document, Dictionary<string, string> propertyMappings)
```

Comment:
Looks like the main purpose of the mapping here is to rename the id property.  Is it possible to rename without mapping everything else?  Might be more efficient?  I know for a JsonObject you'd be able to remove and readd it with a new name.

Reply:
> Looks like the main purpose of the mapping here is to rename the id property.

Not really, it also updates the fields that have custom name from record definition using `StoragePropertyName` property (basically when we need to store `parking_is_included` instead of `ParkingIncluded` and then map it back to C# type for `Get` operations).

I'm also thinking about using mapper in cases where instead of record definition attributes users want to use `BsonElement` attributes, so they can provide a custom mapper that will use these attributes, but I will need to create an additional test and example how to achieve this.

Reply:
Right, that makes sense.